### ISSUE TITLE
better solution for finding the start of "parent expressions"

### ISSUE THREAD

**Issue by oakmac** [oakmac/atom-parinfer#9]
Right now we are assuming that an open paren as the first character of a line is the start of a new expression. There are circumstances where this is not true (example: multi-line strings) and will cause odd behavior from the main Parinfer algorithm.

Need to find a better solution here.

**Comment by jasongilman** [comment]
I'm the maintainer of Proto REPL. I had a similar issue and found that using Atom can tell you the context for any code by using scopes. See https://github.com/jasongilman/proto-repl/blob/master/lib/editor-utils.coffee#L45

**Comment by oakmac** [comment]
Awesome! Thank you for posting this.

**Comment by oakmac** [comment]
[sublime-text-parinfer] has some [interesting discussion] on the "parent expression hack"

[sublime-text-parinfer]:https://github.com/oakmac/sublime-text-parinfer
[interesting discussion]:https://github.com/oakmac/sublime-text-parinfer/issues/23

**Comment by phillc73** [comment]
Just starting out with Clojure (first attempt at any Lisp language) so I hope this is not user error.

Following the "do" example in the [Brace and True Clojure book](http://www.braveclojure.com/do-things/), Parinfer wants to do this:

```Clojure
(if false
  (do (println "Yes! True!")) 
  "True"
  (do (println "No")) 
  "False")
```
Rather than:

```Clojure
(if false
  (do (println "Yes! True!")
  "True!")
  (do (println "No")
  "False"))
```

**Comment by otherjoel** [comment]
> Parinfer in Atom wants to do this [...] rather than this [...]

Yes, this is expected/desired behaviour. Because `"True"` in your example is at the same indent level as the line above it. So Parinfer reasons they must be siblings in the list and adjusts parentheses to reflect this. (same goes for your `"False"` line.)

If you indent `"True"` any further past the indent level of the line above it, Parinfer will see that it is supposed to be a sibling of `(println "Yes! True")` instead of a sibling of `(do ...)` and it will make the parenthses reflect this. Note that this is how the indentation appears in [the original example](http://www.braveclojure.com/do-things/#do) from the book you linked.

The above assumes you're in Indent Mode (where you change indentation manually/directly and Parinfer handles the parens in the background, matching them to the indentation).
